{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd \n",
    "import math\n",
    "import numpy as np \n",
    "from statsmodels.sandbox.stats.multicomp import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initial PATH and Folder Setup\n",
    "PATH='/Users/houy2/Documents/projects/AD/AD_genetics/results/update_v9_Jul072023/MR_results'\n",
    "LD=\"LD02\" #LD clumping threshold: square(R) = 0.2\n",
    "DIR_mr=f'AD_MR_{LD}_Jul172023'\n",
    "DIR_mr_raw=f'AD_MR_{LD}_Jul172023/raw_xQTL'\n",
    "DIR_mr_processed=f'AD_MR_{LD}_Jul172023/processed'\n",
    "DIR_alzMR=f\"AD_MR_{LD}_Jul172023/alzMR_DT\" #the folder include alzMR score.\n",
    "file_type=f\"AD#{LD}\"\n",
    "update_date=\"Jul17\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of MR results across xQTL* AD_GWAS pairs:#3130\n"
     ]
    }
   ],
   "source": [
    "os.chdir(PATH)\n",
    "listname = os.listdir(\"./%s\" % DIR_mr_raw)\n",
    "print(f\"The number of MR results across xQTL* AD_GWAS pairs:#{len(listname)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>Exposure</th>\n",
       "      <th>xQTL_type</th>\n",
       "      <th>method</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>CILower</th>\n",
       "      <th>CIUpper</th>\n",
       "      <th>StdError</th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD_NMAFallNoUKB23_b37_2021N</td>\n",
       "      <td>ZNF408</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "      <td>Inverse variance weighted</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.110594</td>\n",
       "      <td>-0.263294</td>\n",
       "      <td>0.042106</td>\n",
       "      <td>0.077908</td>\n",
       "      <td>0.155740</td>\n",
       "      <td>0.985086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD_NMAFallNoUKB23_b37_2021N</td>\n",
       "      <td>ALMS1P</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "      <td>Inverse variance weighted</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.018612</td>\n",
       "      <td>-0.181386</td>\n",
       "      <td>0.144163</td>\n",
       "      <td>0.083048</td>\n",
       "      <td>0.822674</td>\n",
       "      <td>0.985086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AD_NMAFallNoUKB23_b37_2021N</td>\n",
       "      <td>UBE2R2</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "      <td>Inverse variance weighted</td>\n",
       "      <td>2</td>\n",
       "      <td>0.074615</td>\n",
       "      <td>-0.132360</td>\n",
       "      <td>0.281589</td>\n",
       "      <td>0.105599</td>\n",
       "      <td>0.479825</td>\n",
       "      <td>0.985086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AD_NMAFallNoUKB23_b37_2021N</td>\n",
       "      <td>NACC1</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "      <td>Inverse variance weighted</td>\n",
       "      <td>2</td>\n",
       "      <td>0.070275</td>\n",
       "      <td>-0.050416</td>\n",
       "      <td>0.190965</td>\n",
       "      <td>0.061577</td>\n",
       "      <td>0.253767</td>\n",
       "      <td>0.985086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AD_NMAFallNoUKB23_b37_2021N</td>\n",
       "      <td>DCAF8</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "      <td>Inverse variance weighted</td>\n",
       "      <td>2</td>\n",
       "      <td>0.055664</td>\n",
       "      <td>-0.056125</td>\n",
       "      <td>0.167453</td>\n",
       "      <td>0.057035</td>\n",
       "      <td>0.329084</td>\n",
       "      <td>0.985086</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Outcome Exposure                          xQTL_type   \n",
       "0  AD_NMAFallNoUKB23_b37_2021N   ZNF408  Flead_map_eQTL_brain_mayometa_fdr  \\\n",
       "1  AD_NMAFallNoUKB23_b37_2021N   ALMS1P  Flead_map_eQTL_brain_mayometa_fdr   \n",
       "2  AD_NMAFallNoUKB23_b37_2021N   UBE2R2  Flead_map_eQTL_brain_mayometa_fdr   \n",
       "3  AD_NMAFallNoUKB23_b37_2021N    NACC1  Flead_map_eQTL_brain_mayometa_fdr   \n",
       "4  AD_NMAFallNoUKB23_b37_2021N    DCAF8  Flead_map_eQTL_brain_mayometa_fdr   \n",
       "\n",
       "                      method  SNPs  Estimate   CILower   CIUpper  StdError   \n",
       "0  Inverse variance weighted     2 -0.110594 -0.263294  0.042106  0.077908  \\\n",
       "1  Inverse variance weighted     2 -0.018612 -0.181386  0.144163  0.083048   \n",
       "2  Inverse variance weighted     2  0.074615 -0.132360  0.281589  0.105599   \n",
       "3  Inverse variance weighted     2  0.070275 -0.050416  0.190965  0.061577   \n",
       "4  Inverse variance weighted     2  0.055664 -0.056125  0.167453  0.057035   \n",
       "\n",
       "     Pvalue       FDR  \n",
       "0  0.155740  0.985086  \n",
       "1  0.822674  0.985086  \n",
       "2  0.479825  0.985086  \n",
       "3  0.253767  0.985086  \n",
       "4  0.329084  0.985086  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step1, we combined all MR results by all methods. \n",
    "##Combine all MR results using IVW when the number of IVs =1 OR IV = 1.\n",
    "df_snp2 = pd.DataFrame()\n",
    "for f in listname:\n",
    "    if \"snp\" in f:\n",
    "        df=pd.read_csv(\"./%s/%s\" % (DIR_mr_raw,f), sep=\"\\t\")\n",
    "        df['Outcome']= f.split(\".\")[0].split('#')[1]\n",
    "        xQTL=f.split(\".\")[0].split('#')[2]\n",
    "        df['xQTL_type'] =xQTL\n",
    "        df_snp2 = pd.concat([df, df_snp2])\n",
    "df_snp2=df_snp2[['Outcome','exposure','xQTL_type','method','nsnp','b','lo_ci','up_ci','se','pval','FDR']]\n",
    "df_snp2.rename(columns={'exposure':'Exposure','nsnp':'SNPs','b':'Estimate','se':'StdError','lo_ci':'CILower','up_ci':'CIUpper','pval':'Pvalue'},inplace=True)\n",
    "df_snp2.to_csv(\"./%s/%s_snp2_%s.tsv\" % (DIR_mr_processed,file_type,update_date),index=False,sep='\\t')\n",
    "df_snp2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Exposure</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>method</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>StdError</th>\n",
       "      <th>CILower</th>\n",
       "      <th>CIUpper</th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>Heter.Stat[1]</th>\n",
       "      <th>Heter.p</th>\n",
       "      <th>method_type</th>\n",
       "      <th>FDR</th>\n",
       "      <th>xQTL_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRPEL1</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>17</td>\n",
       "      <td>IVW</td>\n",
       "      <td>-0.006078</td>\n",
       "      <td>0.003373</td>\n",
       "      <td>-0.012690</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>0.071597</td>\n",
       "      <td>20.795037</td>\n",
       "      <td>0.186524</td>\n",
       "      <td>IVW</td>\n",
       "      <td>0.448229</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RP11-529E10.6</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>9</td>\n",
       "      <td>IVW</td>\n",
       "      <td>0.008785</td>\n",
       "      <td>0.005552</td>\n",
       "      <td>-0.002098</td>\n",
       "      <td>0.019667</td>\n",
       "      <td>0.113607</td>\n",
       "      <td>9.456025</td>\n",
       "      <td>0.305297</td>\n",
       "      <td>IVW</td>\n",
       "      <td>0.583433</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTLL5</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>18</td>\n",
       "      <td>IVW</td>\n",
       "      <td>0.002433</td>\n",
       "      <td>0.002713</td>\n",
       "      <td>-0.002885</td>\n",
       "      <td>0.007751</td>\n",
       "      <td>0.369937</td>\n",
       "      <td>19.598959</td>\n",
       "      <td>0.295254</td>\n",
       "      <td>IVW</td>\n",
       "      <td>0.848809</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TDO2</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>15</td>\n",
       "      <td>IVW</td>\n",
       "      <td>-0.002425</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>-0.007721</td>\n",
       "      <td>0.002872</td>\n",
       "      <td>0.369600</td>\n",
       "      <td>11.728905</td>\n",
       "      <td>0.628064</td>\n",
       "      <td>IVW</td>\n",
       "      <td>0.848809</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PHF20L1</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>6</td>\n",
       "      <td>IVW</td>\n",
       "      <td>0.000831</td>\n",
       "      <td>0.007901</td>\n",
       "      <td>-0.014654</td>\n",
       "      <td>0.016316</td>\n",
       "      <td>0.916232</td>\n",
       "      <td>1.850102</td>\n",
       "      <td>0.869473</td>\n",
       "      <td>IVW</td>\n",
       "      <td>0.995892</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Exposure          Outcome  SNPs method  Estimate  StdError   CILower   \n",
       "0         GRPEL1  AD_Jansen_2019N    17    IVW -0.006078  0.003373 -0.012690  \\\n",
       "1  RP11-529E10.6  AD_Jansen_2019N     9    IVW  0.008785  0.005552 -0.002098   \n",
       "2          TTLL5  AD_Jansen_2019N    18    IVW  0.002433  0.002713 -0.002885   \n",
       "3           TDO2  AD_Jansen_2019N    15    IVW -0.002425  0.002702 -0.007721   \n",
       "4        PHF20L1  AD_Jansen_2019N     6    IVW  0.000831  0.007901 -0.014654   \n",
       "\n",
       "    CIUpper    Pvalue  Heter.Stat[1]   Heter.p method_type       FDR   \n",
       "0  0.000534  0.071597      20.795037  0.186524         IVW  0.448229  \\\n",
       "1  0.019667  0.113607       9.456025  0.305297         IVW  0.583433   \n",
       "2  0.007751  0.369937      19.598959  0.295254         IVW  0.848809   \n",
       "3  0.002872  0.369600      11.728905  0.628064         IVW  0.848809   \n",
       "4  0.016316  0.916232       1.850102  0.869473         IVW  0.995892   \n",
       "\n",
       "                           xQTL_type  \n",
       "0  Flead_map_eQTL_brain_mayometa_fdr  \n",
       "1  Flead_map_eQTL_brain_mayometa_fdr  \n",
       "2  Flead_map_eQTL_brain_mayometa_fdr  \n",
       "3  Flead_map_eQTL_brain_mayometa_fdr  \n",
       "4  Flead_map_eQTL_brain_mayometa_fdr  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Combine all MR results using IVW when the number of IVs >=3.\n",
    "df_IVW = pd.DataFrame()\n",
    "for f in listname:\n",
    "    if \"IVW\" in f:\n",
    "        df=pd.read_csv(\"./%s/%s\" % (DIR_mr_raw,f), sep=\"\\t\")\n",
    "        df['Outcome']= f.split(\".\")[0].split('#')[1]\n",
    "        xQTL=f.split(\".\")[0].split('#')[2]\n",
    "        df['xQTL_type'] =xQTL\n",
    "        df_IVW = pd.concat([df, df_IVW])\n",
    "df_IVW.to_csv(\"./%s/%s_IVW3_%s.tsv\" % (DIR_mr_processed,file_type,update_date),index=False,sep='\\t')\n",
    "df_IVW.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(576902, 14)\n"
     ]
    }
   ],
   "source": [
    "#Combined results from the Wald ratio and IVW.\n",
    "df_Wald_IVW = pd.concat([df_snp2, df_IVW])\n",
    "df_Wald_IVW =df_Wald_IVW.dropna(subset=[\"Exposure\"])\n",
    "df_Wald_IVW.to_csv(\"./%s/%s_IVW_%s.tsv\" % (DIR_mr_processed,file_type,update_date),index=False,sep='\\t')\n",
    "print(df_Wald_IVW.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combined results from the Maxlik, Weighted median and MRpresso.\n",
    "df_Maxlik = pd.DataFrame()\n",
    "df_WM = pd.DataFrame()\n",
    "df_MP = pd.DataFrame()\n",
    "for f in listname:\n",
    "    #print(f)\n",
    "    if \"Maxlik#\" in f:\n",
    "        df=pd.read_csv(\"./%s/%s\" % (DIR_mr_raw,f), sep=\"\\t\")\n",
    "        df['Outcome']= f.split(\".\")[0].split('#')[1]\n",
    "        xQTL=f.split(\".\")[0].split('#')[2]\n",
    "        df['xQTL_type'] =xQTL\n",
    "        df = df.dropna(subset=['Exposure','FDR']).copy()\n",
    "        df_Maxlik = pd.concat([df, df_Maxlik])\n",
    "    elif \"WM#\" in f:\n",
    "        df=pd.read_csv(\"./%s/%s\" % (DIR_mr_raw,f), sep=\"\\t\")\n",
    "        df['Outcome']= f.split(\".\")[0].split('#')[1]\n",
    "        xQTL=f.split(\".\")[0].split('#')[2]\n",
    "        df['xQTL_type'] =xQTL\n",
    "        df = df.dropna(subset=['Exposure','FDR']).copy()\n",
    "        df_WM = pd.concat([df, df_WM])\n",
    "    elif \"MRpresso#\" in f:\n",
    "        df=pd.read_csv(\"./%s/%s\" % (DIR_mr_raw,f), sep=\"\\t\")\n",
    "        df = df.drop(['outcome','T-stat','xqtl'], axis=1).copy()\n",
    "        df['Outcome']= f.split(\".\")[0].split('#')[1]\n",
    "        xQTL=f.split(\".\")[0].split('#')[2]\n",
    "        df['xQTL_type'] =xQTL\n",
    "        df_drop = df.dropna(subset=['Causal Estimate','Sd','P-value']).copy()\n",
    "        df_drop=df_drop.sort_values(by=[\"MR Analysis\"]).drop_duplicates(subset=['Exposure']).reset_index()\n",
    "        p_adjusted = multipletests(df_drop[\"P-value\"].tolist(), alpha=0.05, method='fdr_bh', is_sorted=False, returnsorted=False)[1]\n",
    "        df_drop['FDR']=p_adjusted\n",
    "        df_drop['method']=\"MRpresso\"\n",
    "        df_MP = pd.concat([df_drop, df_MP])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Exposure</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>method</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>StdError</th>\n",
       "      <th>CILower</th>\n",
       "      <th>CIUpper</th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>Heter.Stat[1]</th>\n",
       "      <th>Heter.p</th>\n",
       "      <th>FDR</th>\n",
       "      <th>method_type</th>\n",
       "      <th>xQTL_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTSO</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>11</td>\n",
       "      <td>MaxLik</td>\n",
       "      <td>-0.005085</td>\n",
       "      <td>0.002658</td>\n",
       "      <td>-0.010295</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.055785</td>\n",
       "      <td>9.093810</td>\n",
       "      <td>0.523224</td>\n",
       "      <td>0.417923</td>\n",
       "      <td>maxlik</td>\n",
       "      <td>Flead_map_eQTL_brain_rosmap_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>THAP9-AS1</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>4</td>\n",
       "      <td>MaxLik</td>\n",
       "      <td>-0.002743</td>\n",
       "      <td>0.002410</td>\n",
       "      <td>-0.007467</td>\n",
       "      <td>0.001980</td>\n",
       "      <td>0.254999</td>\n",
       "      <td>1.911446</td>\n",
       "      <td>0.590988</td>\n",
       "      <td>0.689707</td>\n",
       "      <td>maxlik</td>\n",
       "      <td>Flead_map_eQTL_brain_rosmap_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TET2</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>6</td>\n",
       "      <td>MaxLik</td>\n",
       "      <td>0.001995</td>\n",
       "      <td>0.004340</td>\n",
       "      <td>-0.006512</td>\n",
       "      <td>0.010501</td>\n",
       "      <td>0.645814</td>\n",
       "      <td>7.087317</td>\n",
       "      <td>0.214227</td>\n",
       "      <td>0.887790</td>\n",
       "      <td>maxlik</td>\n",
       "      <td>Flead_map_eQTL_brain_rosmap_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>REST</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>3</td>\n",
       "      <td>MaxLik</td>\n",
       "      <td>-0.003133</td>\n",
       "      <td>0.005653</td>\n",
       "      <td>-0.014213</td>\n",
       "      <td>0.007946</td>\n",
       "      <td>0.579379</td>\n",
       "      <td>0.064183</td>\n",
       "      <td>0.968418</td>\n",
       "      <td>0.853859</td>\n",
       "      <td>maxlik</td>\n",
       "      <td>Flead_map_eQTL_brain_rosmap_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>POLN</td>\n",
       "      <td>AD_Jansen_2019N</td>\n",
       "      <td>3</td>\n",
       "      <td>MaxLik</td>\n",
       "      <td>0.004713</td>\n",
       "      <td>0.007032</td>\n",
       "      <td>-0.009069</td>\n",
       "      <td>0.018496</td>\n",
       "      <td>0.502693</td>\n",
       "      <td>2.709089</td>\n",
       "      <td>0.258065</td>\n",
       "      <td>0.835806</td>\n",
       "      <td>maxlik</td>\n",
       "      <td>Flead_map_eQTL_brain_rosmap_fdr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Exposure          Outcome  SNPs  method  Estimate  StdError   CILower   \n",
       "0       CTSO  AD_Jansen_2019N    11  MaxLik -0.005085  0.002658 -0.010295  \\\n",
       "1  THAP9-AS1  AD_Jansen_2019N     4  MaxLik -0.002743  0.002410 -0.007467   \n",
       "2       TET2  AD_Jansen_2019N     6  MaxLik  0.001995  0.004340 -0.006512   \n",
       "3       REST  AD_Jansen_2019N     3  MaxLik -0.003133  0.005653 -0.014213   \n",
       "4       POLN  AD_Jansen_2019N     3  MaxLik  0.004713  0.007032 -0.009069   \n",
       "\n",
       "    CIUpper    Pvalue  Heter.Stat[1]   Heter.p       FDR method_type   \n",
       "0  0.000126  0.055785       9.093810  0.523224  0.417923      maxlik  \\\n",
       "1  0.001980  0.254999       1.911446  0.590988  0.689707      maxlik   \n",
       "2  0.010501  0.645814       7.087317  0.214227  0.887790      maxlik   \n",
       "3  0.007946  0.579379       0.064183  0.968418  0.853859      maxlik   \n",
       "4  0.018496  0.502693       2.709089  0.258065  0.835806      maxlik   \n",
       "\n",
       "                         xQTL_type  \n",
       "0  Flead_map_eQTL_brain_rosmap_fdr  \n",
       "1  Flead_map_eQTL_brain_rosmap_fdr  \n",
       "2  Flead_map_eQTL_brain_rosmap_fdr  \n",
       "3  Flead_map_eQTL_brain_rosmap_fdr  \n",
       "4  Flead_map_eQTL_brain_rosmap_fdr  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Maxlik.to_csv(\"./%s/%s_MaxLik_%s.tsv\" % (DIR_mr_processed,file_type,update_date),index=False,sep='\\t')\n",
    "df_Maxlik.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Exposure</th>\n",
       "      <th>method_type</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>StdError</th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>xQTL_type</th>\n",
       "      <th>FDR</th>\n",
       "      <th>method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>45</td>\n",
       "      <td>ZNF813</td>\n",
       "      <td>Outlier-corrected</td>\n",
       "      <td>-0.043925</td>\n",
       "      <td>0.025868</td>\n",
       "      <td>0.231597</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>Flead_map_eQTL_brain_mayocer_fdr</td>\n",
       "      <td>0.517233</td>\n",
       "      <td>MRpresso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>83</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>Outlier-corrected</td>\n",
       "      <td>0.011748</td>\n",
       "      <td>0.029241</td>\n",
       "      <td>0.714761</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>Flead_map_eQTL_brain_mayocer_fdr</td>\n",
       "      <td>0.845100</td>\n",
       "      <td>MRpresso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>367</td>\n",
       "      <td>PCDHB14</td>\n",
       "      <td>Outlier-corrected</td>\n",
       "      <td>-0.043900</td>\n",
       "      <td>0.021961</td>\n",
       "      <td>0.295295</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>Flead_map_eQTL_brain_mayocer_fdr</td>\n",
       "      <td>0.568641</td>\n",
       "      <td>MRpresso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>GUCA1B</td>\n",
       "      <td>Raw</td>\n",
       "      <td>-0.014213</td>\n",
       "      <td>0.014127</td>\n",
       "      <td>0.343818</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>Flead_map_eQTL_brain_mayocer_fdr</td>\n",
       "      <td>0.597056</td>\n",
       "      <td>MRpresso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>258</td>\n",
       "      <td>PPAP2B</td>\n",
       "      <td>Raw</td>\n",
       "      <td>-0.018472</td>\n",
       "      <td>0.027712</td>\n",
       "      <td>0.541525</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>Flead_map_eQTL_brain_mayocer_fdr</td>\n",
       "      <td>0.751855</td>\n",
       "      <td>MRpresso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  Exposure        method_type  Estimate  StdError    Pvalue   \n",
       "0     45    ZNF813  Outlier-corrected -0.043925  0.025868  0.231597  \\\n",
       "1     83  C1orf112  Outlier-corrected  0.011748  0.029241  0.714761   \n",
       "2    367   PCDHB14  Outlier-corrected -0.043900  0.021961  0.295295   \n",
       "3      0    GUCA1B                Raw -0.014213  0.014127  0.343818   \n",
       "4    258    PPAP2B                Raw -0.018472  0.027712  0.541525   \n",
       "\n",
       "                  Outcome                         xQTL_type       FDR   \n",
       "0  AD_Kunkle_Stage1_2019N  Flead_map_eQTL_brain_mayocer_fdr  0.517233  \\\n",
       "1  AD_Kunkle_Stage1_2019N  Flead_map_eQTL_brain_mayocer_fdr  0.845100   \n",
       "2  AD_Kunkle_Stage1_2019N  Flead_map_eQTL_brain_mayocer_fdr  0.568641   \n",
       "3  AD_Kunkle_Stage1_2019N  Flead_map_eQTL_brain_mayocer_fdr  0.597056   \n",
       "4  AD_Kunkle_Stage1_2019N  Flead_map_eQTL_brain_mayocer_fdr  0.751855   \n",
       "\n",
       "     method  \n",
       "0  MRpresso  \n",
       "1  MRpresso  \n",
       "2  MRpresso  \n",
       "3  MRpresso  \n",
       "4  MRpresso  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_MP.rename(columns={'Sd':'StdError','P-value':'Pvalue','Causal Estimate':'Estimate',\n",
    "                            \"MR Analysis\":\"method_type\"},inplace=True)\n",
    "df_MP.to_csv(\"./%s/%s_MRpresso_%s.tsv\" % (DIR_mr_processed,file_type,update_date),index=False,sep='\\t')\n",
    "df_MP.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Exposure</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>method</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>StdError</th>\n",
       "      <th>CILower</th>\n",
       "      <th>CIUpper</th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>method_type</th>\n",
       "      <th>iterations</th>\n",
       "      <th>xQTL_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GNRHR</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>15</td>\n",
       "      <td>WeightedMedian</td>\n",
       "      <td>-0.070216</td>\n",
       "      <td>0.026397</td>\n",
       "      <td>-0.121952</td>\n",
       "      <td>-0.018479</td>\n",
       "      <td>0.007814</td>\n",
       "      <td>0.356919</td>\n",
       "      <td>Weighted_Median</td>\n",
       "      <td>10000</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCRG1</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>44</td>\n",
       "      <td>WeightedMedian</td>\n",
       "      <td>-0.011429</td>\n",
       "      <td>0.018172</td>\n",
       "      <td>-0.047045</td>\n",
       "      <td>0.024187</td>\n",
       "      <td>0.529383</td>\n",
       "      <td>0.934970</td>\n",
       "      <td>Weighted_Median</td>\n",
       "      <td>10000</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZCRB1</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>19</td>\n",
       "      <td>WeightedMedian</td>\n",
       "      <td>-0.001228</td>\n",
       "      <td>0.032358</td>\n",
       "      <td>-0.064649</td>\n",
       "      <td>0.062193</td>\n",
       "      <td>0.969729</td>\n",
       "      <td>0.993078</td>\n",
       "      <td>Weighted_Median</td>\n",
       "      <td>10000</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RP6-91H8.5</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>14</td>\n",
       "      <td>WeightedMedian</td>\n",
       "      <td>0.042106</td>\n",
       "      <td>0.031094</td>\n",
       "      <td>-0.018837</td>\n",
       "      <td>0.103048</td>\n",
       "      <td>0.175686</td>\n",
       "      <td>0.821960</td>\n",
       "      <td>Weighted_Median</td>\n",
       "      <td>10000</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RP11-806K15.1</td>\n",
       "      <td>AD_Kunkle_Stage1_2019N</td>\n",
       "      <td>12</td>\n",
       "      <td>WeightedMedian</td>\n",
       "      <td>0.020580</td>\n",
       "      <td>0.034908</td>\n",
       "      <td>-0.047839</td>\n",
       "      <td>0.088999</td>\n",
       "      <td>0.555496</td>\n",
       "      <td>0.946045</td>\n",
       "      <td>Weighted_Median</td>\n",
       "      <td>10000</td>\n",
       "      <td>Flead_map_eQTL_brain_mayometa_fdr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Exposure                 Outcome  SNPs          method  Estimate   \n",
       "0          GNRHR  AD_Kunkle_Stage1_2019N    15  WeightedMedian -0.070216  \\\n",
       "1          SCRG1  AD_Kunkle_Stage1_2019N    44  WeightedMedian -0.011429   \n",
       "2          ZCRB1  AD_Kunkle_Stage1_2019N    19  WeightedMedian -0.001228   \n",
       "3     RP6-91H8.5  AD_Kunkle_Stage1_2019N    14  WeightedMedian  0.042106   \n",
       "4  RP11-806K15.1  AD_Kunkle_Stage1_2019N    12  WeightedMedian  0.020580   \n",
       "\n",
       "   StdError   CILower   CIUpper    Pvalue       FDR      method_type   \n",
       "0  0.026397 -0.121952 -0.018479  0.007814  0.356919  Weighted_Median  \\\n",
       "1  0.018172 -0.047045  0.024187  0.529383  0.934970  Weighted_Median   \n",
       "2  0.032358 -0.064649  0.062193  0.969729  0.993078  Weighted_Median   \n",
       "3  0.031094 -0.018837  0.103048  0.175686  0.821960  Weighted_Median   \n",
       "4  0.034908 -0.047839  0.088999  0.555496  0.946045  Weighted_Median   \n",
       "\n",
       "   iterations                          xQTL_type  \n",
       "0       10000  Flead_map_eQTL_brain_mayometa_fdr  \n",
       "1       10000  Flead_map_eQTL_brain_mayometa_fdr  \n",
       "2       10000  Flead_map_eQTL_brain_mayometa_fdr  \n",
       "3       10000  Flead_map_eQTL_brain_mayometa_fdr  \n",
       "4       10000  Flead_map_eQTL_brain_mayometa_fdr  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_WM.to_csv(\"./%s/%s_WM_%s.tsv\" % (DIR_mr_processed,file_type,update_date),index=False,sep='\\t')\n",
    "df_WM.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1096930, 16)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Combined all MR reuslts by different methods.\n",
    "df_all = pd.concat([df_Wald_IVW,df_Maxlik,df_WM,df_MP])\n",
    "df_all.to_csv(\"./%s/%s_MRall_%s_#%s.tsv\" % (DIR_mr_processed,file_type,update_date,df_all.shape[0]),index=False,sep='\\t')\n",
    "df_all.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Step2, re-organized files names of GWAS and xQTL data.\n",
    "reps = {\n",
    "    'Flead-eQTL_metabrain_BasalGanglia':'eQTL_Metabrain_BasalGanglia',\n",
    "    'Flead-eQTL_metabrain_Cerebellum':'eQTL_Metabrain_Cerebellum',\n",
    "    'Flead-eQTL_metabrain_CortexAFA':'eQTL_Metabrain_CortexAFA',\n",
    "    'Flead-eQTL_metabrain_Cortex':'eQTL_Metabrain_Cortex',\n",
    "    'Flead-eQTL_metabrain_Hippocampus':'eQTL_Metabrain_Hippocampus',\n",
    "    'Flead-eQTL_metabrain_SpinalCord':'eQTL_Metabrain_SpinalCord',\n",
    "    'Flead_map_eQTL_brain_mayocer_fdr':'eQTL_Mayo_Cerebellum',\n",
    "    'Flead_map_eQTL_brain_mayometa_fdr':'eQTL_Mayo_MetaAnalysis',\n",
    "    'Flead_map_eQTL_brain_mayotcx_fdr':'eQTL_Mayo_Cortex',\n",
    "    'Flead_map_eQTL_brain_rosmap_fdr':'eQTL_ROSMAP_Cortex',\n",
    "    'Flead_map_pQTL_brain_banner_fdr':'pQTL_Banner_Cortex',\n",
    "    'Flead_map_pQTL_brain_rosmapall_fdr':'pQTL_ROSMAP_Cortex.All',\n",
    "    'Flead_map_pQTL_brain_rosmapcontrol_fdr':'pQTL_ROSMAP_Cortex.Ctl',\n",
    "    'AD_Bellenguez_2022N':'AD_Bellenguez_2022',\n",
    "    'AD_NMAFallNoUKB23_b37_2021N':'AD.NoUKB23_Wightman_2021',\n",
    "    'AD_Kunkle_AFA_M2_b37_2020N':'AD.AAsAPOE_Kunkle_2020',\n",
    "    'AD_Kunkle_AFA_M1_b37_2020N':'AD.AAs_Kunkle_2020',\n",
    "    'AD_Jansen_2019N':'AD_Jansen_2019',\n",
    "    'AD_GCST90012878_b37_2021N':'AD.FamilyHtx_Schwartzentruber_2021',\n",
    "    'AD_GCST90012877_b37_2021N':'AD_Schwartzentruber_2021',\n",
    "    'AD_Kunkle_Stage2_2019N':'AD.Stage2_Kunkle_2019',\n",
    "    'AD_Kunkle_Stage1_2019N':'AD.Stage1_Kunkle_2019',\n",
    "    'Inverse variance weighted':'IVW',\n",
    "    'Wald ratio':'WaldRatio'\n",
    "    }\n",
    "\n",
    "def replace_all(text, reps_dic):\n",
    "    for i, j in reps_dic.items():\n",
    "        text = text.replace(i, j)\n",
    "    return text\n",
    "\n",
    "listname_mr = os.listdir(\"./%s\" % DIR_mr_processed)\n",
    "for filename in listname_mr:\n",
    "    if \".tsv\" in filename:\n",
    "        with open(\"./%s/%s\" % (DIR_mr_processed,filename),'r') as f:\n",
    "            text=f.read()\n",
    "            text=replace_all(text,reps)\n",
    "        \n",
    "        with open(\"./%s/processed-%s\" % (DIR_mr_processed,filename),'w') as w:\n",
    "            w.write(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processed-AD#LD02_IVW3_Jul17#192106\n",
      "processed-AD#LD02_snp2_Jul17#361082\n",
      "processed-AD#LD02_WM_Jul17#192098\n",
      "processed-AD#LD02_IVW_Jul17#552983\n",
      "processed-AD#LD02_MRall_Jul17_#1096930#1008269\n",
      "processed-AD#LD02_MaxLik_Jul17#192098\n",
      "processed-AD#LD02_MRpresso_Jul17#135732\n"
     ]
    }
   ],
   "source": [
    "#Combine MR results from Kunkle.Stage1 and Stage2 to Stage1&2\n",
    "listname_mr = os.listdir(\"./%s\" % DIR_mr_processed)\n",
    "for filename in listname_mr:\n",
    "    if \"processed-\" in filename:\n",
    "        df_processed=pd.read_csv(f\"./{DIR_mr_processed}/{filename}\", sep=\"\\t\", low_memory=False)\n",
    "        \n",
    "        df_KunkleStage = df_processed[\n",
    "            ((df_processed['Outcome'] == \"AD.Stage1_Kunkle_2019\") | (df_processed['Outcome'] == \"AD.Stage2_Kunkle_2019\")) & \n",
    "            (df_processed['xQTL_type'] != \"eQTL_Metabrain_CortexAFA\") \n",
    "        ]\n",
    "        df_KunkleStage = df_KunkleStage.sort_values(by=['Pvalue'], ascending=True).drop_duplicates(subset= ['Exposure', 'xQTL_type'])\n",
    "        df_KunkleStage.loc[:, 'Outcome'] = df_KunkleStage['Outcome'].str.replace(\"Stage1\", \"Stage1&2\")\n",
    "        df_KunkleStage.loc[:, 'Outcome'] = df_KunkleStage['Outcome'].str.replace(\"Stage2\", \"Stage1&2\")        \n",
    "        df_others=df_processed[(df_processed['Outcome'] != \"AD.Stage1_Kunkle_2019\") & \n",
    "                               (df_processed['Outcome'] != \"AD.Stage2_Kunkle_2019\") & \n",
    "                               (df_processed['xQTL_type'] != \"eQTL_Metabrain_CortexAFA\")]\n",
    "        df_processed_update = pd.concat([df_KunkleStage,df_others]).reset_index(drop=True)\n",
    "        name = filename.split(\".\")[0]\n",
    "        print(f\"{name}#{df_processed_update.shape[0]}\")\n",
    "        df_processed_update.to_csv(f\"./{DIR_mr_processed}/update_{filename}\", sep='\\t',index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count significant MR results across all xQTL* AD_GWAS pairs.\n",
    "sig#: n-m, where n represents the number of significant MR results with β-coefficient > 0 at FDR < 0.05 in EA (European American) and P < 0.05 in AA (African American); m represents the number of significant MR results with β-coefficient < 0 at FDR < 0.05 in EA or P < 0.05 in AA. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique(list1): \n",
    "    x = np.array(list1) \n",
    "    return set(np.unique(x))\n",
    "def count_posi(k,test_list):\n",
    "    count = len([i for i in test_list if i > k])\n",
    "    return count\n",
    "def count_neg(k,test_list):\n",
    "    count = len([i for i in test_list if i < k])\n",
    "    return count\n",
    "def count_var(k,test_list):\n",
    "    count = len([i for i in test_list if i == k])\n",
    "    return count\n",
    "\n",
    "def MR_count(df,MR_type,sig_type,cutoff,sample_type,update_time,PATH_out):\n",
    "    #generate a MR score and save in PATH_out dir.\n",
    "    df_count = df.groupby(['Exposure']).count()\n",
    "    df_count=df_count.sort_values(['method'], ascending=False).rename(columns={'method':'total#'})\n",
    "    print(f\"How many drug targets for {sample_type} at {sig_type}< 0.05:#{df_count.shape[0]}\")\n",
    "    df_count_sub=df_count[[\"total#\"]].copy()\n",
    "    df_sig=df[df[sig_type]<0.05]\n",
    "    sig_gene_list=unique(df_sig[\"Exposure\"].tolist())\n",
    "    for gene in df_count_sub.index:\n",
    "        gene_sub_all=df[df['Exposure']==gene]\n",
    "        EST_list_all=gene_sub_all['Estimate'].astype(float).tolist()\n",
    "        posi_count_all=count_posi(0,EST_list_all)\n",
    "        df_count_sub.loc[gene,\"posi#\"]=posi_count_all\n",
    "        neg_count_all=count_neg(0,EST_list_all)\n",
    "        df_count_sub.loc[gene,\"neg#\"]=neg_count_all\n",
    "        if gene in sig_gene_list:\n",
    "            gene_sub=df_sig[df_sig['Exposure']==gene]\n",
    "            EST_list=gene_sub['Estimate'].astype(float).tolist()\n",
    "            n=count_posi(0,EST_list)\n",
    "            df_count_sub.loc[gene,\"n.#sig_posi\"]=int(n)\n",
    "            m=count_neg(0,EST_list)\n",
    "            df_count_sub.loc[gene,\"m.#sig_neg\"]=int(m)\n",
    "            df_count_sub.loc[gene,\"sig#\"]=n-m\n",
    "            df_count_sub.loc[gene,\"abs_sig#\"]=n+m\n",
    "\n",
    "    df_count_sub=df_count_sub.reset_index().rename(columns={\"Exposure\":\"symbol\"})\n",
    "    print(df_count_sub.head())\n",
    "    df_count_sub.to_csv(\"%s/%s_%s%s_%s_score_%s_sig.tsv\" % (PATH_out,MR_type,sig_type,cutoff,sample_type,update_time), sep=\"\\t\", index=False)\n",
    "    return df_count_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38627\n",
      "=======end======\n"
     ]
    }
   ],
   "source": [
    "#Filter the MR results for Drug Targets.\n",
    "df_MR_IVW = pd.read_csv(f\"./{DIR_mr_processed}/update_processed-{file_type}_IVW_{update_date}.tsv\",sep=\"\\t\",low_memory=False)\n",
    "df_target = pd.read_csv(f\"./target_#1415.tsv\",sep=\"\\t\")\n",
    "df_target.rename(columns={'symbol': \"Exposure\"},inplace = True)\n",
    "df_DT = pd.merge(df_MR_IVW,df_target,how = \"inner\",on =[\"Exposure\"])\n",
    "print(f\"{df_DT.shape[0]}\")\n",
    "print(\"=======end======\")\n",
    "df_DT.to_csv(f\"./{DIR_mr}/DT-{file_type}_IVW_{update_date}_#{df_DT.shape[0]}.tsv\", sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "How many drug targets for DT_EUR at FDR< 0.05:#1229\n",
      "  symbol  total#  posi#  neg#  n.#sig_posi  m.#sig_neg  sig#  abs_sig#\n",
      "0  EPHX2      72   70.0   2.0         22.0         0.0  22.0      22.0\n",
      "1   PEPD      72    8.0  64.0          0.0         2.0  -2.0       2.0\n",
      "2    GAA      66   33.0  33.0          2.0         0.0   2.0       2.0\n",
      "3   NQO2      66   28.0  38.0          NaN         NaN   NaN       NaN\n",
      "4  SIRT5      66   20.0  46.0          NaN         NaN   NaN       NaN\n",
      "How many drug targets for DT_EUR at Pvalue< 0.05:#1229\n",
      "  symbol  total#  posi#  neg#  n.#sig_posi  m.#sig_neg  sig#  abs_sig#\n",
      "0  EPHX2      72   70.0   2.0         40.0         1.0  39.0      41.0\n",
      "1   PEPD      72    8.0  64.0          0.0        14.0 -14.0      14.0\n",
      "2    GAA      66   33.0  33.0          6.0         4.0   2.0      10.0\n",
      "3   NQO2      66   28.0  38.0          2.0         0.0   2.0       2.0\n",
      "4  SIRT5      66   20.0  46.0          0.0         1.0  -1.0       1.0\n",
      "===============\n",
      "How many drug targets for DT_AA at FDR< 0.05:#1222\n",
      "  symbol  total#  posi#  neg#  n.#sig_posi  m.#sig_neg  sig#  abs_sig#\n",
      "0   PEPD      24    9.0  15.0          NaN         NaN   NaN       NaN\n",
      "1  EPHX2      24    8.0  16.0          NaN         NaN   NaN       NaN\n",
      "2    GAA      22   12.0  10.0          NaN         NaN   NaN       NaN\n",
      "3  SIRT5      22    3.0  19.0          NaN         NaN   NaN       NaN\n",
      "4   NQO2      22   18.0   4.0          NaN         NaN   NaN       NaN\n",
      "How many drug targets for DT_AA at Pvalue< 0.05:#1222\n",
      "  symbol  total#  posi#  neg#  n.#sig_posi  m.#sig_neg  sig#  abs_sig#\n",
      "0   PEPD      24    9.0  15.0          NaN         NaN   NaN       NaN\n",
      "1  EPHX2      24    8.0  16.0          0.0         3.0  -3.0       3.0\n",
      "2    GAA      22   12.0  10.0          0.0         1.0  -1.0       1.0\n",
      "3  SIRT5      22    3.0  19.0          NaN         NaN   NaN       NaN\n",
      "4   NQO2      22   18.0   4.0          NaN         NaN   NaN       NaN\n"
     ]
    }
   ],
   "source": [
    "#Calculating the number of significant for Drug Targets\n",
    "#Drug targets are defined based on a binding affinity less than 1000nM between the protein and ligand.\n",
    "\n",
    "#EA-specific\n",
    "df_EUR=df_DT[(df_DT['Outcome'] != 'AD.AAsAPOE_Kunkle_2020') & (df_DT['Outcome'] != \"AD.AAs_Kunkle_2020\")].copy()\n",
    "#print(f\"The number of MR results for EA: #{df_EUR.shape[0]}\")\n",
    "df_EUR.to_csv(f\"./{DIR_mr_processed}/update_processedEUR-{file_type}_IVW_{update_date}.tsv\", sep='\\t',index=False)\n",
    "EUR_q=MR_count(df_EUR,file_type,'FDR',0.05,'DT_EUR',update_date,DIR_alzMR)\n",
    "EUR_p=MR_count(df_EUR,file_type,'Pvalue',0.05,'DT_EUR',update_date,DIR_alzMR)\n",
    "print(\"===============\")\n",
    "#AA-specific\n",
    "df_AA=df_DT[(df_DT['Outcome'] == 'AD.AAsAPOE_Kunkle_2020') | \n",
    "                               (df_DT['Outcome'] == \"AD.AAs_Kunkle_2020\") ].copy()\n",
    "#print(f\"The number of MR results for AA: #{df_AA.shape[0]}\")\n",
    "df_AA.to_csv(f\"./{DIR_mr_processed}/update_processedAA-{file_type}_IVW_{update_date}.tsv\", sep='\\t',index=False)\n",
    "AA_q=MR_count(df_AA,file_type,'FDR',0.05,'DT_AA',update_date,DIR_alzMR)\n",
    "AA_p=MR_count(df_AA,file_type,'Pvalue',0.05,'DT_AA',update_date,DIR_alzMR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>abs_sig#_AA.P</th>\n",
       "      <th>AA-EA_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>SETD7</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ALDH2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>SGK1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>TRPV3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    symbol  abs_sig#_AA.P  AA-EA_ratio\n",
       "331  SETD7            5.0          5.0\n",
       "274   ABL1            4.0          4.0\n",
       "44   ALDH2            4.0          4.0\n",
       "565   SGK1            4.0          4.0\n",
       "104  TRPV3            6.0          4.0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def preprocess_df(df, suffix):\n",
    "    df = df.add_suffix(f'_{suffix}')\n",
    "    columns_to_keep = [f\"symbol_{suffix}\", f\"n.#sig_posi_{suffix}\", f\"m.#sig_neg_{suffix}\", f\"sig#_{suffix}\", f\"abs_sig#_{suffix}\"]\n",
    "    df = df[columns_to_keep].rename(columns={f\"symbol_{suffix}\":\"symbol\"})\n",
    "    return df\n",
    "\n",
    "df_AAp = preprocess_df(AA_p, 'AA.P')\n",
    "df_EAp = preprocess_df(EUR_p, 'EA.P')\n",
    "df_AAp.fillna(0, inplace=True)\n",
    "df_EAp.fillna(0, inplace=True)\n",
    "\n",
    "df_AA_EA = pd.merge(df_AAp, df_EAp, how=\"outer\", on=\"symbol\")\n",
    "df_AA_EA['AA-EA_ratio'] = (df_AA_EA['abs_sig#_AA.P'] - df_AA_EA['abs_sig#_EA.P'])\n",
    "df_AA_EA = df_AA_EA.sort_values(by='AA-EA_ratio', ascending=False)\n",
    "df_AA_EA_ratio = df_AA_EA[df_AA_EA['AA-EA_ratio'] >= 2][['symbol', 'abs_sig#_AA.P', 'AA-EA_ratio']]\n",
    "#df_AA_EA_ratio.to_csv(f\"./{DIR_alzMR}/alzScores_DT_EA_AA_ratio.tsv\", index=False, sep=\"\\t\")\n",
    "df_AA_EA_ratio.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## alzMR-Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function for alzMR-Scores\n",
    "def calculate_alzMRscores(data):\n",
    "    # Create score for each gene\n",
    "    score = data.groupby('Exposure').agg(\n",
    "        K=('Z', 'size'),\n",
    "        SG=('Z', 'sum'),\n",
    "        meanBeta=('Estimate', 'mean'),\n",
    "        meanPosBeta=('Estimate', lambda x: x[x > 0].mean() if not x[x > 0].empty else 0),\n",
    "        meanNegBeta=('Estimate', lambda x: x[x < 0].mean() if not x[x < 0].empty else 0),\n",
    "        nPosBeta=('Estimate', lambda x: (x > 0).sum()),\n",
    "        nNegBeta=('Estimate', lambda x: (x < 0).sum())\n",
    "    ).reset_index()\n",
    "\n",
    "    score['SGK'] = score['SG'] / score['K']\n",
    "\n",
    "    # Rankings\n",
    "    score = score.reindex(score.SG.abs().sort_values(ascending=False).index)\n",
    "    score['rankSumZscores'] = np.arange(1, len(score) + 1)\n",
    "    score = score.reindex(score.SGK.abs().sort_values(ascending=False).index)\n",
    "    score['rankMeanZscores'] = np.arange(1, len(score) + 1)\n",
    "    score = score.sort_values(by='rankSumZscores')\n",
    "\n",
    "    score = score[['Exposure', 'SG', 'SGK', 'rankSumZscores', 'rankMeanZscores']]\n",
    "    score.columns = ['symbol', 'alzMRScore', 'Nor-alzMRScore', 'rankSumZscores', \n",
    "                     'rankMeanZscores']\n",
    "    score.reset_index(drop=True,inplace=True)\n",
    "    return score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>alzMRScore</th>\n",
       "      <th>Nor-alzMRScore</th>\n",
       "      <th>rankSumZscores</th>\n",
       "      <th>rankMeanZscores</th>\n",
       "      <th>total#</th>\n",
       "      <th>posi#</th>\n",
       "      <th>neg#</th>\n",
       "      <th>n.#sig_posi</th>\n",
       "      <th>m.#sig_neg</th>\n",
       "      <th>sig#</th>\n",
       "      <th>abs_sig#</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACE</td>\n",
       "      <td>-199.231645</td>\n",
       "      <td>-4.150659</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EPHX2</td>\n",
       "      <td>195.698354</td>\n",
       "      <td>2.718033</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>72</td>\n",
       "      <td>70.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LACTB</td>\n",
       "      <td>165.475084</td>\n",
       "      <td>3.064353</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>54</td>\n",
       "      <td>54.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADAM10</td>\n",
       "      <td>133.563373</td>\n",
       "      <td>3.180080</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>42</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERCC2</td>\n",
       "      <td>126.195368</td>\n",
       "      <td>3.505427</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>36</td>\n",
       "      <td>33.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   symbol  alzMRScore  Nor-alzMRScore  rankSumZscores  rankMeanZscores   \n",
       "0     ACE -199.231645       -4.150659               1                2  \\\n",
       "1   EPHX2  195.698354        2.718033               2               10   \n",
       "2   LACTB  165.475084        3.064353               3                9   \n",
       "3  ADAM10  133.563373        3.180080               4                7   \n",
       "4   ERCC2  126.195368        3.505427               5                5   \n",
       "\n",
       "   total#  posi#  neg#  n.#sig_posi  m.#sig_neg  sig#  abs_sig#  \n",
       "0      48    0.0  48.0          0.0        32.0 -32.0      32.0  \n",
       "1      72   70.0   2.0         22.0         0.0  22.0      22.0  \n",
       "2      54   54.0   0.0         21.0         0.0  21.0      21.0  \n",
       "3      42   42.0   0.0         17.0         0.0  17.0      17.0  \n",
       "4      36   33.0   3.0         20.0         0.0  20.0      20.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate Z-statistics from beta/se\n",
    "df_EUR['Z'] = df_EUR['Estimate'] / df_EUR['StdError']\n",
    "MRscore = calculate_alzMRscores(df_EUR)\n",
    "MRscore=pd.merge(MRscore,EUR_q, how=\"left\",on=\"symbol\")\n",
    "\n",
    "MRscore.to_csv(f\"./{DIR_alzMR}/alzScores_DT_EUR_Zstatistics.tsv\", index=False, sep=\"\\t\")\n",
    "MRscore.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>alzMRScore</th>\n",
       "      <th>Nor-alzMRScore</th>\n",
       "      <th>rankSumZscores</th>\n",
       "      <th>rankMeanZscores</th>\n",
       "      <th>abs_sig#_AA.P</th>\n",
       "      <th>AA-EA_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EBPL</td>\n",
       "      <td>-30.367974</td>\n",
       "      <td>-1.897998</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR</td>\n",
       "      <td>24.016763</td>\n",
       "      <td>1.501048</td>\n",
       "      <td>4</td>\n",
       "      <td>41</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DHFR</td>\n",
       "      <td>21.116581</td>\n",
       "      <td>1.173143</td>\n",
       "      <td>7</td>\n",
       "      <td>111</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALDH2</td>\n",
       "      <td>-20.740114</td>\n",
       "      <td>-1.296257</td>\n",
       "      <td>8</td>\n",
       "      <td>72</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SLC7A11</td>\n",
       "      <td>20.707722</td>\n",
       "      <td>1.294233</td>\n",
       "      <td>9</td>\n",
       "      <td>73</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    symbol  alzMRScore  Nor-alzMRScore  rankSumZscores  rankMeanZscores   \n",
       "0     EBPL  -30.367974       -1.897998               1               12  \\\n",
       "1      BCR   24.016763        1.501048               4               41   \n",
       "2     DHFR   21.116581        1.173143               7              111   \n",
       "3    ALDH2  -20.740114       -1.296257               8               72   \n",
       "4  SLC7A11   20.707722        1.294233               9               73   \n",
       "\n",
       "   abs_sig#_AA.P  AA-EA_ratio  \n",
       "0            6.0          2.0  \n",
       "1            4.0          2.0  \n",
       "2            3.0          2.0  \n",
       "3            4.0          4.0  \n",
       "4            3.0          2.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate Z-statistics from beta/se\n",
    "df_AA['Z'] = df_AA['Estimate'] / df_AA['StdError']\n",
    "MRscore_AA = calculate_alzMRscores(df_AA)\n",
    "MRscore_AA=pd.merge(MRscore_AA,df_AA_EA_ratio, how=\"inner\",on=\"symbol\")\n",
    "MRscore_AA.to_csv(f\"./{DIR_alzMR}/alzScores_DT_AA_Zstatistics.tsv\", index=False, sep=\"\\t\")\n",
    "MRscore_AA.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
